vividus-extension-selenium/src/main/java/org/vividus/ui/listener/AbstractSourceCodePublishingOnFailureListener.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2019-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.vividus.ui.listener;
+
+import java.util.Map;
+
+import com.google.common.eventbus.Subscribe;
+
+import org.vividus.reporter.event.IAttachmentPublisher;
+import org.vividus.selenium.IWebDriverProvider;
+import org.vividus.softassert.event.AssertionFailedEvent;
+
+public abstract class AbstractSourceCodePublishingOnFailureListener
+{
+    private final IAttachmentPublisher attachmentPublisher;
+    private final IWebDriverProvider webDriverProvider;
+
+    private boolean publishSourcesOnFailure;
+
+    protected AbstractSourceCodePublishingOnFailureListener(IAttachmentPublisher attachmentPublisher,
+            IWebDriverProvider webDriverProvider)
+    {
+        this.webDriverProvider = webDriverProvider;
+        this.attachmentPublisher = attachmentPublisher;
+    }
+
+    @Subscribe
+    public void onAssertionFailure(AssertionFailedEvent event)
+    {
+        if (publishSourcesOnFailure && webDriverProvider.isWebDriverInitialized())
+        {
+            attachmentPublisher.publishAttachment("/templates/source-code.ftl",
+                    Map.of("sourceCode", getSourceCode(), "format", getFormat()), "Application source code");
+        }
+    }
+
+    protected abstract String getSourceCode();
+
+    protected abstract String getFormat();
```

Comment:
it's better to have a constructor parameter instead

Reply:
why? by returning this static info from the method we reduce amount of configurations